Mathlib/Analysis/Complex/ValueDistribution/CountingFunction.lean

```diff
@@ -0,0 +1,217 @@
+    (D : locallyFinsuppWithin (univ : Set E) ℤ) :
+    logCounting D 0 = 0 := by
+  rw [logCounting]
+  simp
+
+end Function.locallyFinsuppWithin
+
+/-!
+## The Logarithmic Counting Function of a Meromorphic Function
+-/
+
+namespace ValueDistribution
+
+variable
+  {𝕜 : Type*} [NontriviallyNormedField 𝕜] [ProperSpace 𝕜]
+  {E : Type*} [NormedAddCommGroup E] [NormedSpace 𝕜 E] [CompleteSpace E]
+  {U : Set 𝕜} {f g : 𝕜 → E} {a : WithTop E} {a₀ : E}
+
+variable (f a) in
+/--
+The logarithmic counting function of a meromorphic function.
+
+If `f : 𝕜 → E` is meromorphic and `a : WithTop E` is any value, this is a logarithmically weighted
+measure of the number of times the function `f` takes a given value `a` within the disk `∣z∣ ≤ r`,
+counting multiplicities.  In the special case where `a = ⊤`, it counts the poles of `f`.
+-/
+noncomputable def logCounting : ℝ → ℝ := by
+  by_cases h : a = ⊤
+  · exact (divisor f univ)⁻.logCounting
+  · exact (divisor (fun z ↦ f z - a.untop₀) univ)⁺.logCounting
+
+/--
+For finite values `a₀`, the logarithmic counting function `logCounting f a₀` is the counting
+function associated with the zero-divisor of the meromorphic function `f - a₀`.
+-/
+lemma logCounting_finite :
+    logCounting f a₀ = (divisor (fun z ↦ f z - a₀) univ)⁺.logCounting := by
+  simp [logCounting]
+
+/--
+For finite values `a₀`, the logarithmic counting function `logCounting f a₀` equals the logarithmic
+counting function for the value zero of the shifted function `f - a₀`.
+-/
+lemma logCounting_finite_eq_logCounting_zero_of_shifted :
+    logCounting f a₀ = logCounting (f - fun _ ↦ a₀) 0 := by
+  simp [logCounting]
+
+/--
+The logarithmic counting function `logCounting f 0` is the counting function associated with the
+zero-divisor of `f`.
+-/
+lemma logCounting_zero :
+    logCounting f 0 = (divisor f univ)⁺.logCounting := by
+  simp [logCounting]
+
+/--
+The logarithmic counting function `logCounting f Set.univ` is the counting function associated with
+the pole-divisor of `f`.
+-/
+lemma logCounting_top :
+    logCounting f ⊤ = (divisor f univ)⁻.logCounting := by
+  simp [logCounting]
+
+/--
+Evaluation of the logarithmic counting function at zero yields zero.
+-/
+lemma logCounting_eval_zero :
+    logCounting f a 0 = 0 := by
+  by_cases h : a = ⊤ <;> simp [logCounting, h]
+
+/--
+The counting function associated with the divisor of `f` is the difference between `logCounting f 0`
+and `logCounting f ⊤`.
+-/
+theorem log_counting_zero_sub_logCounting_top {f : 𝕜 → E} :
+    (divisor f univ).logCounting = logCounting f 0 - logCounting f ⊤ := by
+  rw [← posPart_sub_negPart (divisor f univ), logCounting_zero, logCounting_top, map_sub]
+
+/-!
+## Elementary Properties of the Counting Function
+-/
+
+/--
+Relation between the logarithmic counting function of `f` and of `f⁻¹`.
+-/
+theorem logCounting_inv {f : 𝕜 → 𝕜} :
+    logCounting f 0 = logCounting f⁻¹ ⊤ := by
```

Comment:
```suggestion
     logCounting f⁻¹ ⊤ = logCounting f 0 := by
```
In general, put the most complicated term to the left, as this is the natural direction for `rw` and `simp`.